include/quic/connection_ids.hpp

```diff
@@ -0,0 +1,108 @@
+        ReferenceID(ReferenceID&& obj) = default;
+
+        ReferenceID& operator=(const ReferenceID& obj) = default;
+        ReferenceID& operator=(ReferenceID&& obj) = default;
+
+        inline bool operator<(const ReferenceID& other) const { return id < other.id; }
+
+        explicit operator const uint64_t&() const { return id; }
+        // operator uint64_t&() { return id; }
+
+        std::string to_string() const { return "< RID:{} >"_format(id); }
+    };
+    template <>
+    constexpr inline bool IsToStringFormattable<ReferenceID> = true;
+
+    // Wrapper for ngtcp2_cid with helper functionalities to make it passable
+    struct alignas(size_t) ConnectionID final : ngtcp2_cid
+    {
+        ConnectionID() = default;
+        ConnectionID(const ConnectionID& c) = default;
+        ConnectionID(ngtcp2_cid c) : ConnectionID(c.data, c.datalen) {}
+        ConnectionID(const uint8_t* cid, size_t length)
+        {
+            assert(length <= NGTCP2_MAX_CIDLEN);
+            datalen = length;
+            std::memmove(data, cid, datalen);
+        }
+
+        ConnectionID& operator=(const ConnectionID& c) = default;
+
+        inline bool operator==(const ConnectionID& other) const
+        {
+            return datalen == other.datalen && std::memcmp(data, other.data, datalen) == 0;
+        }
+
+        inline bool operator!=(const ConnectionID& other) const { return !(*this == other); }
+
+        static ConnectionID random()
+        {
+            ConnectionID cid;
+            cid.datalen = static_cast<size_t>(NGTCP2_MAX_CIDLEN);
+            gnutls_rnd(GNUTLS_RND_RANDOM, cid.data, cid.datalen);
+            return cid;
+        }
+
+        std::string to_string() const
+        {
+            return "{:02x}"_format(fmt::join(std::begin(data), std::begin(data) + datalen, ""));
+        }
+    };
+    template <>
+    constexpr inline bool IsToStringFormattable<ConnectionID> = true;
+}  // namespace oxen::quic
+
+namespace std
+{
+    // Custom hash is required s.t. unordered_set storing ConnectionID:unique_ptr<Connection>
+    // is able to call its implicit constructor
+    template <>
+    struct hash<oxen::quic::ConnectionID>
+    {
+        size_t operator()(const oxen::quic::ConnectionID& cid) const
+        {
+            static_assert(
+                    alignof(oxen::quic::ConnectionID) >= alignof(size_t) &&
+                    offsetof(oxen::quic::ConnectionID, data) % sizeof(size_t) == 0);
+            return *reinterpret_cast<const size_t*>(cid.data);
+        }
+    };
+
+    template <>
+    struct hash<oxen::quic::ReferenceID>
+    {
+        size_t operator()(const oxen::quic::ReferenceID& rid) const
+        {
+            static_assert(
+                    alignof(oxen::quic::ReferenceID) >= alignof(uint64_t) &&
+                    offsetof(oxen::quic::ReferenceID, id) % sizeof(uint64_t) == 0);
+            return *reinterpret_cast<const size_t*>(rid.id);
```

Comment:
Since the reference ID is just a wrapped integer value there's no need for the reinterpret_cast on the data (which is only really something you want for random data -- which is fine for ConnectionID, but not great for a monotonically increasing value).  Instead you can just defer this one to the built-in integer hasher:
```suggestion
            return std::hash<decltype(rid.id)>{}(rid.id);
```